### PR TITLE
Target Encoder: add initial code and unit test

### PR DESCRIPTION
**Pull request by fsywang**
Addresses #25 #33

### CODE REVIEW

File: encoderpy/target_encoder.py [hunk 1]
```diff
@@ -1,47 +1,67 @@
...
+        if isinstance(cat_columns, list) == False:
+                raise Exception("Type of cat_columns must be a list")
+        if (isinstance(prior, float) | isinstance(prior, int)) == False:
+                raise Exception("Type of prior must be a numeric value")
+        if (isinstance(X_train, pd.DataFrame) & isinstance(X_test, pd.DataFrame)) == False:
+                raise Exception("Type of X_train and X_test must be pd.Dataframe")
+        if isinstance(y, pd.Series) == False:
+                raise Exception("Type of y must be pd.Series")
+        if (set(cat_columns).issubset(X_train.columns) & set(cat_columns).issubset(X_test.columns)) == False:
+                raise Exception("X_train and X_test must contain cat_columns")
```
Comment (braydentang1) [inline encoderpy/target_encoder.py]: Awesome checks, but maybe a couple suggestions:

- If the user inputs y as a pd.Series that contains characters, will your code break? Perhaps we can implement an if statement that automatically encodes the target as a 0-1 indicator variable, so that we can take means when the user does this.
- If the user inputs y as a pd.Series with more than two unique values and specify as a binary problem, will the code break?
- Also, don't forget to allow for a null test set as in #42 .

File: encoderpy/target_encoder.py [hunk 2]
```diff
@@ -1,47 +1,67 @@
...
+        test_processed = X_test.copy()
+
+        for col in cat_columns:
+                # calculate target mean for each category and save to dictionary
+                search_table = train_processed.groupby(col)[y.name].mean().to_dict()
+                # encode categorical columns for training dataset
+                train_processed.loc[:,col] = train_processed[col].map(search_table)
+                # encode categorical columns for testing dataset
+                test_processed.loc[:,col] = test_processed[col].map(search_table)
+                test_processed.loc[:,col] = test_processed[col].fillna(prior)
```
Comment (braydentang1) [inline encoderpy/target_encoder.py]: Are we still implementing the Laplace smoothing? This is completely correct by the way, but the prior should also impact the training set encodings as well since they get weighted by the global mean. Here is some code that I think works, please tell me what you think/feel free to change whatever you want!

```suggestion
                 global_mean = np.mean(y)
                # calculate target counts for each category and save to dictionary
                search_table = train_processed.groupby(col)[y.name].aggregate(['sum', 'count'])
                
                search_table['encodings'] = (search_table['sum'] + prior * global_mean) / (search_table['count'] + prior)
                search_table = search_table.drop(columns=['sum', 'count'], axis=1).to_dict()['encodings']
                
                # encode categorical columns for training dataset
                train_processed.loc[:,col] = train_processed[col].map(search_table)
                # encode categorical columns for testing dataset
                test_processed.loc[:,col] = test_processed[col].map(search_table)
                test_processed.loc[:,col] = test_processed[col].fillna(prior)  
```

File: tests/test_target_encoder.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+def test_output():
+    train_encode1, test_encode1 = target_encoder(train1,test1,
+                                                 train.target_bin, 
+                                                 ['feature_cat_chr','feature_cat_num'])
+    train_encode2, test_encode2 = target_encoder(train2,test2,
+                                                 train.target_bin, 
+                                                 ['feature_cat_chr','feature_cat_num'], prior = 0.1)
+    #test value
+    assert test_encode2['feature_cat_chr'].iloc[0] == 0.1, 'The encoded value for unseen test dataset should equal to prior'
+    assert np.isclose(train_encode1['feature_cat_chr'].iloc[0], 0.4285714) == True,'The encoded value for training dataset is wrong'
```
Comment (braydentang1) [inline tests/test_target_encoder.py]: Nice! I know you are still working on this so there's not much to say, but if we decide to implement the Laplace smoothing then the check here will change (the encoding value will be influenced by the global mean).

File: encoderpy/target_encoder.py [hunk 4]
```diff
@@ -1,47 +1,67 @@
...
-          in the test set. A larger value gives less weight to what is observed in the training set. A value
-          of 0 incorporates no prior information. The default value is 0.5.
-  min_samples: int
-          The minimum samples to calculate mean of targets. If number of the target smaller than min_samples, 
-          the value will be encoded as prior probability. The default value is 1.
-          
-  Returns
-  -------
-  train_processed : pd.DataFrame
+def target_encoder(X_train, X_test, y, cat_columns, prior = 0.5):
```
Comment (braydentang1) [inline encoderpy/target_encoder.py]: I think we need to import the packages this function depends on. So pandas and numpy?

**Review by braydentang1 — CHANGES_REQUESTED** [review]
Nice job! I just added some comments and suggestions, feel free to disagree/change as you see fit!

File: encoderpy/target_encoder.py (hunk 4 above)
Comment (fsywang) [reply encoderpy/target_encoder.py]: I added import lines in the new commit. But is it necessary to add the import in the function script?

**Comment by fsywang** [comment]
@braydentang1 Thank you for the feedback! The new commit above resolves them.  For the new file, I did the following:
For code:
- Update docstring and input argument
- Add more input edge cases
- Change the smoothing formula
-  Set X_test to default None and handle both cases #42 

For test:
- Covers all if statement in the function

**Review by braydentang1 — APPROVED** [review]
Thanks for the changes!
